### DISCUSSION TITLE
React useMemo with observable state

### DISCUSSION THREAD

**Discussion by kykungz** [mobxjs/mobx#3348] [Q&A] [answered]
Hi, I'm wondering what's the correct way to use React's `useMemo` with MobX observable and React's props/state as its dependencies.

Considering a simple example such as having a `threshold` value in a MobX Store, which will be used as a setting to filter out numbers that is less than the threshold number.

```ts
class Store {
  threshold = 5

  constructor() {
    makeAutoObservable(this)
  }
}
```

I saw in the docs about [Tip: useEffect and observables](https://mobx.js.org/react-integration.html#useeffect), which has an example of using `useEffect` with observables like this:
```tsx
// Run the effect when `store.threshold` changed.
useEffect(
  () =>
    autorun(() => {
      const list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
      const filtered = list.filter(value => value > store.threshold)
      alert(filtered)
    }),
  []
)
```

Or you can also combine values from React props/state with MobX observable in `useEffect`:
```tsx
// Run the effect when `store.threshold` or `props.list` changed.
useEffect(
  () =>
    autorun(() => {
      const filtered = props.list.filter(value => value > store.threshold)
      alert(filtered)
    }),
  [props.list]
)
```
---
So I'm wondering what's the correct way to do the same thing, but as a memoized value instead of side effects.

```tsx
// Memoizing value when `store.threshold` changed
const filtered = useMemo(() => {
  const list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
  return list.filter(value => value > store.threshold)
}, [store.threshold])
```

The above code shows eslint warning as follows:
> React Hook useMemo has an unnecessary dependency: 'store.threshold'. Either exclude it or remove the dependency array. Outer scope values like 'store.threshold' aren't valid dependencies because mutating them doesn't re-render the component.
>
> eslint([react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920))

Then I found a way to correctly do it without eslint warning using [computed](https://mobx.js.org/computeds-with-args.html#2-close-over-the-arguments) function that behaves like a `useMemo` for observables:
```tsx
// Caching (memoizing) value when `store.threshold` changed
const filtered = computed(() => {
  const list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
  return list.filter(value => value > store.threshold)
}).get()
```

But when it comes to listening to both `props.list` and `store.threshold`, I have no idea how to do it without errors or eslint warning.
```tsx
// ❌ eslint warning react-hooks/exhaustive-deps

// Memoizing value when `props.list` or `store.threshold` changed
const filtered = useMemo(() => {
  return props.list.filter(value => value > store.threshold)
}, [props.list, store.threshold])
```
```tsx
// ❌ changes in `store.threshold` will not be reflected

// Memoizing value when `props.list` or `store.threshold` changed
const filtered = useMemo(() => {
  return computed(() => props.list.filter(value => value > store.threshold)).get()
}, [props.list])
```
```tsx
// ❌ changes in `props.list` will not be reflected

// Memoizing value when `props.list` or `store.threshold` changed
const filtered = computed(() => {
  return props.list.filter(value => value > store.threshold)
}).get()
```

Is there a correct way to do this without breaking the eslint rules? Or my ultimate solution is to disable the eslint rule with:

```tsx
// Memoizing value when `props.list` or `store.threshold` changed
const filtered = useMemo(() => {
  return props.list.filter(value => value > store.threshold)
// eslint-disable-next-line react-hooks/exhaustive-deps
}, [props.list, store.threshold])
```

**Comment by kykungz** [comment, marked as answer]
Just found out a solution when tweaking around with the code.

Instead of returning the final value through `.get()` inside `useMemo`
```tsx
// ❌ changes in `store.threshold` will not be reflected
const filtered = useMemo(() => {
  return computed(() => props.list.filter(value => value > store.threshold)).get()
}, [props.list])
```
Just return the everything as a computed value from `useMemo` and use `.get()` afterwards:
```tsx
// ✅ changes in `props.list` AND `store.threshold` will now be reflected
const filtered = useMemo(() => {
  return computed(() => props.list.filter(value => value > store.threshold))
}, [props.list]).get()
```
`store.threshold` will be tracked through `computed` function, `props.list` will be tracked through `useMemo`. Both changes will be returned as a computed value from `computed`, to be called `.get()` at the end.

**Comment by urugator** [comment]
Memoization + mutable state is indeed tricky:
https://github.com/mobxjs/mobx/discussions/3252
https://github.com/mobxjs/mobx/discussions/3224

**Reply by kykungz** [reply]
Code in the both discussions are using MobX observable as dependencies of React hooks (`useMemo`, `useEffect`),
which indeed would cause unexpected behavior due to mutable nature of MobX.

But on the above answer, immutable React props/states are handled through React hooks, while MobX mutable state are handled through MobX's `computed`, which would not cause a problem... isn't it?

**Comment by kubk** [comment]
I'd suggest not to mix props & Mobx at all. Either do your calculations in Mobx and use its `computed` or use React immutable model (props, state)

**Reply by kykungz** [reply]
It is a common situation where we need to mix MobX states with React states isn't it? Like for an example, having user settings store on MobX, which will be used to apply to each component's state.

It is also mentioned in the official docs on how to mix React props in `useEffect` and use `autorun` at the same time. So I think it is safe to do so(?)
